Split Lambda troubleshooting into its own page (#6781)
DataDog/documentation: content/en/logs/guide/lambda-logs-collection-troubleshooting-guide.md

---
title: Lambda Function Log Collection Troubleshooting Guide
kind: documentation
---

> ## Logs from lambda functions are not visible in Log Explorer page

Reviewer: ```suggestion
## Logs from Lambda functions are not visible in Log Explorer page
```

Reviewer 2: I would actually remove this title as it is kind of a duplicate of the main article title. WDYT?

Author: What about getting rid of the title and changing the line after it to:

>If you don't see logs forwarded from a lambda function in the Log Explorer, see the Datadog-AWS Log integration to configure your environment.

Reviewer 2: Sounds like a good plan indeed

See the [Datadog-AWS Log integration][1] to configure your environment. If you still do not see your logs, double-check the following points:

> There are some common issues that can come up when configuring the AWS Lambda function to forward logs to Datadog. Below are some troubleshooting steps to try if you are using the Lambda function and your logs aren't reaching Datadog. If you continue to have trouble after following these steps, [contact our support team][2] for further assistance.

Reviewer: ```suggestion
There are some common issues that can come up when configuring the AWS Lambda function to forward logs to Datadog. Below are some troubleshooting steps to try if you are using the Lambda function and your logs aren't reaching Datadog. If you continue to have trouble after following these steps, [contact support][2] for further assistance.
```

> ## Check if logs are reaching the Live Tail

Reviewer 2: Let's phrase it different, as a question.

Something like, `Are your logs sent to Datadog?` which is actually what we check with the livetail

1. Navigate to the [Log Explorer's Live Tail View][3].
2. In the Search Bar, use a filter to limit the Live Tail View to just the logs coming from your lambda function. Some common search queries are:
    * By source: the source is often set to `source:aws` or `source:cloudwatch` but you can find other possible sources in the `parse_event_source` function in the [Lambda function][4]. 
    * By forwarder name: the Lambda function adds a forwardername tag to all the logs it forwards. You can filter on this tag by searching for `forwardername:*` or `forwardername:<function_name>`.
3. If you see your logs in the Live Tail View but they are not reaching the Log Explorer, this could be due to one of the following reasons:

>     * Logs are too old: we drop logs with a timestamp older than 6 hours.

Reviewer 2: ```suggestion
    * Logs are too far in the past: Datadog only accepts logs with a timestamp older than 6 hours in the past or 1h in the future.
```

    * If your index has any [exclusion filters][5] set up, they may be filtering out your logs.

> ## Check the Lambda function monitoring tab

Reviewer 2: What do you think of something like `The lambda function is not triggered`

From the Lambda Function: 

1. Click the Monitoring tab.

{{< img src="logs/guide/lambda-monitoring-tab.png" alt="Monitoring tab"  style="width:80%;" >}}

2. The monitoring tab displays a series of graphs indicating the following information about your Lambda function: 
    * invocations
    * errors
    * logs

> 3. If you don't see any data points on the "Invocations" graph, there may be a problem with the triggers you set for your function. See [Manage Your Function Triggers](#manage-your-function-triggers). To get insight into your lambda invocations without using the monitoring tab, see [Viewing Lambda metrics in Datadog](#viewing-lambda-metrics-in-datadog).

Reviewer: ```suggestion
3. If you don't see any data points on the **Invocations** graph, there may be a problem with the triggers you set for your function. See [Manage Your Function Triggers](#manage-your-function-triggers). To get insight into your lambda invocations without using the monitoring tab, see [Viewing Lambda metrics in Datadog](#viewing-lambda-metrics-in-datadog).
```

4. If you see data points on the "Error count and success rate" graph, [Check the Lambda function logs](#check-the-lambda-function-logs) to see what error messages are being reported.

> ## Viewing Lambda metrics in Datadog

Reviewer 2: Should this be more of a note ? or a ### rather than a title as the same level as the others?

If you have enabled AWS Lambda metrics, you can view metrics related to Lambda invocations and errors within Datadog. The following metrics are all tagged with a key value `functionname` tag: 

> * `aws.lambda.invocations ` : Count of times the Lambda function has been triggered / invoked

Reviewer: ```suggestion
* `aws.lambda.invocations `: Count of times the Lambda function has been triggered / invoked
```

> * `aws.lambda.errors ` : Count of errors that occurred when the function was invoked

Reviewer: ```suggestion
* `aws.lambda.errors `: Count of errors that occurred when the function was invoked
```

> * `aws.lambda.duration ` : Average amount of time (in milliseconds) that it took for the lambda function to finish executing

Reviewer: ```suggestion
* `aws.lambda.duration ` : Average amount of time (in milliseconds) that it took for the Lambda function to finish executing
```

> * `aws.lambda.duration ` : Average amount of time (in milliseconds) that it took for the lambda function to finish executing

Reviewer: ```suggestion
* `aws.lambda.duration `: Average amount of time (in milliseconds) that it took for the lambda function to finish executing
```

> * `aws.lambda.duration.maximum` : Maximum amount of time (in milliseconds) that it took for the lambda function to finish executing

Reviewer: ```suggestion
* `aws.lambda.duration.maximum`: Maximum amount of time (in milliseconds) that it took for the Lambda function to finish executing
```

> * `aws.lambda.throttles` : Count of invocation attempts that were throttled due to invocation rates exceeding customer limits

Reviewer: ```suggestion
* `aws.lambda.throttles`: Count of invocation attempts that were throttled due to invocation rates exceeding customer limits
```

> For more information on these and other AWS lambda metrics, see [Amazon Lambda Metrics][6].

Reviewer: ```suggestion
For more information on these and other AWS Lambda metrics, see [Amazon Lambda Metrics][6].
```

> ## Check the Lambda function logs

Reviewer 2: What do you think of `Configuration issue` ?

Author: Hmm. My reasoning was to present it more as a set of tools for where people can look to check for indicators of possible issues (monitors, logs). Then the sub-sections after that list some possible issues you could find in each of those places.

In this case they're prompted to check for errors in the logs so they know where to go as a first step when gathering information before creating a support ticket. The API key being misconfigured is just one example of something you might spot from the logs.

Reviewer 2: I leave it up to you. It's the idea of do we want people to see sections has potential things to do or to identify issues that they might have.

Both are valid.

Author: Sounds good. I'll go ahead and leave it as-is for now but we can change based on feedback if necessary. (The above comments apply to line 20 as well).

> From the monitoring tab, click "View logs in Cloudwatch".

Reviewer: ```suggestion
From the monitoring tab, click **View logs in Cloudwatch**.
```

{{< img src="logs/guide/lambda-logs-cloudwatch.png" alt="Lambda logs in Cloudwatch"  style="width:80%;" >}}

### API key

If you see one of the following log lines it means that your API key isn't set correctly.

```
module initialization error: Missing Datadog API key
```
This means you didn't set the API key. Places you can set the API key are:
* Directly in the Python code of your Lambda function
* As an environment variable under the name `DD_API_KEY`
* As a KMS Encrypted key under the name `DD_KMS_API_KEY`

```
module initialization error: The API key is not valid.
```
The API key you provided doesn't match any key Datadog recognizes. Double-check the spelling and make sure that your key corresponds to the org you are sending data to. If your org is in Europe, you need to specify a `DD_SITE` variable with the value `datadoghq.eu` in order for the API key to be properly matched with your account.

```
module initialization error: The API key is not the expected length. Please confirm that your API key is correct
```
The API key is either too short or too long. Double-check that you copied it over correctly.

> ## Manage your function triggers

Reviewer 2: Seems to be redundant with the above section where we ask to check the monitoring tab.

Should we package those together?

### Check that the function has triggers configured (manual)

You can see if there are [manual triggers][7] set up for your Lambda function by looking directly in the Lambda function's Configuration tab as in the screenshot below:

{{< img src="logs/guide/manual-triggers-example.png" alt="Example of manual triggers location"  style="width:80%;" >}}

**Note** If you have triggers on your Lambda function but it still isn't being invoked, there may be a conflict with another resource already subscribed to the same log source. When you add a manual trigger, an error message informs you if a resource is already subscribed to the log source:

{{< img src="logs/guide/creating-trigger-error-example.png" alt="Example of error when creating trigger with subscription"  style="width:80%;" >}}

See [Check for conflicting subscriptions](#check-for-conflicting-subscriptions) for more information on removing subscriptions.

### Check that the function has triggers configured (automatic)

You can see if there are [automatic triggers][8] set up for your Lambda function with the following steps:

1. Navigate to the [Cloudwatch console](https://console.aws.amazon.com/cloudwatch/).

> 2. Click "Log Groups" in the left sidebar) here you will see a list of Log Groups. There is a "subscriptions" column on the right that shows what resources (if any) are currently subscribed to the log source.

Reviewer: ```suggestion
2. Click **Log Groups** in the left sidebar here you will see a list of Log Groups. There is a **subscriptions** column on the right that shows what resources (if any) are currently subscribed to the log source.
```

3. If your Lambda function isn't listed as the subscriber for the Log Group you want to monitor, redo the steps from the [automatic trigger setup documentation][8].
4. If the Log Group you want to monitor already has a different resource subscribed to it, see [Check for conflicting subscriptions](#check-for-conflicting-subscriptions) below.

## Check for conflicting subscriptions

AWS doesn't allow for more than one resource to be subscribed to a log source. If you have triggers on your Lambda function but it still isn't being invoked, there may be a conflict with another resource already subscribed to the same log source. The log source needs to be freed up so the Datadog Lambda function can collect logs from it.

If something is already subscribing to a Log Group that you want to monitor with the Datadog Lambda Function, you can remove it: 
* Select the log source 

> * Select "Remove Subscription Filter" in the "Actions" pulldown

Reviewer: ```suggestion
* Select **Remove Subscription Filter** in the **Actions** pulldown
```

---
title: Lambda Function Log Collection Troubleshooting Guide
kind: documentation
---

If you don't see logs forwarded from a lambda function in the Log Explorer, see the [Datadog-AWS Log integration][1] to configure your environment. If you still do not see your logs, double-check the following points:

> There are some common issues that can come up when configuring the AWS Lambda function to forward logs to Datadog. Below are some troubleshooting steps to try if you are using the Lambda function and your logs aren't reaching Datadog. If you continue to have trouble after following these steps, [contact support][2] for further assistance.

Reviewer 3: ```suggestion
There are some common issues that can come up when configuring the AWS Lambda function to forward logs to Datadog. Below are some troubleshooting steps to try if you are using the Lambda function and your logs aren't reaching Datadog. If you continue to have trouble after following these steps, [contact Datadog support][2] for further assistance.
```

## Are your logs sent to Datadog?

1. Navigate to the [Log Explorer's Live Tail View][3].
2. In the Search Bar, use a filter to limit the Live Tail View to just the logs coming from your lambda function. Some common search queries are:
    * By source: the source is often set to `source:aws` or `source:cloudwatch` but you can find other possible sources in the `parse_event_source` function in the [Lambda function][4]. 

>     * By forwarder name: the Lambda function adds a forwardername tag to all the logs it forwards. You can filter on this tag by searching for `forwardername:*` or `forwardername:<function_name>`.

Reviewer 3: ```suggestion
    * By forwarder name: the Lambda function adds a forwardername tag to all the logs it forwards. You can filter on this tag by searching for `forwardername:*` or `forwardername:<FUNCTION_NAME>`.
```

3. If you see your logs in the Live Tail View but they are not reaching the Log Explorer, this could be due to one of the following reasons:
    * Logs are too far in the past: Datadog only accepts logs with a timestamp older than 6 hours in the past or 1h in the future.
    * If your index has any [exclusion filters][5] set up, they may be filtering out your logs.

> 

Reviewer 2: I would add a transition sentence here, like:

 If you don't see the logs in the livetail, the logs are not reaching Datadog so let's investigate the most common configuration issues.

Or something similar

## Check the Lambda function monitoring tab

> From the Lambda Function: 

Reviewer 2: Can we put `from the AWS console` with a link to the Lambda service?

1. Click the Monitoring tab.

> {{< img src="logs/guide/lambda-monitoring-tab.png" alt="Monitoring tab"  style="width:80%;" >}}

Reviewer 3: ```suggestion
    {{< img src="logs/guide/lambda-monitoring-tab.png" alt="Monitoring tab"  style="width:80%;" >}}
```

2. The monitoring tab displays a series of graphs indicating the following information about your Lambda function: 

>     * invocations

Reviewer 3: ```suggestion
    
    * invocations
```

    * errors
    * logs

3. If you don't see any data points on the **Invocations** graph, there may be a problem with the triggers you set for your function. See [Manage Your Function Triggers](#manage-your-function-triggers). To get insight into your lambda invocations without using the monitoring tab, see [Viewing Lambda metrics in Datadog](#viewing-lambda-metrics-in-datadog).
4. If you see data points on the "Error count and success rate" graph, [Check the Lambda function logs](#check-the-lambda-function-logs) to see what error messages are being reported.

### Viewing Lambda metrics in Datadog

> If you have enabled AWS Lambda metrics, you can view metrics related to Lambda invocations and errors within Datadog. The following metrics are all tagged with a key value `functionname` tag: 

Reviewer 3: ```suggestion
If you have enabled AWS Lambda metrics, you can view metrics related to Lambda invocations and errors within Datadog. The following metrics are all tagged with the `functionname` tag: 
```

> * `aws.lambda.invocations `: Count of times the Lambda function has been triggered / invoked

Reviewer 3: We should make an array of this for clarity

1. Navigate to the [Cloudwatch console](https://console.aws.amazon.com/cloudwatch/).
2. Click **Log Groups** in the left sidebar here you will see a list of Log Groups. There is a **subscriptions** column on the right that shows what resources (if any) are currently subscribed to the log source.
3. If your Lambda function isn't listed as the subscriber for the Log Group you want to monitor, redo the steps from the [automatic trigger setup documentation][1].
4. If the Log Group you want to monitor already has a different resource subscribed to it, see [Check for conflicting subscriptions](#check-for-conflicting-subscriptions) below.

[1]: https://docs.datadoghq.com/integrations/amazon_web_services/?tab=allpermissions#automatically-setup-triggers
{{% /tab %}}
{{< /tabs >}}

## Check the Lambda function logs

From the monitoring tab, click **View logs in Cloudwatch**.

{{< img src="logs/guide/lambda-logs-cloudwatch.png" alt="Lambda logs in Cloudwatch"  style="width:80%;" >}}

### API key

If you see one of the following log lines it means that your API key isn't set correctly.

```
module initialization error: Missing Datadog API key
```

> This means you didn't set the API key. Places you can set the API key are:

Reviewer 3: ```suggestion

This means you didn't set the API key. Places you can set the API key are:
```

> * Directly in the Python code of your Lambda function

Reviewer 3: ```suggestion

* Directly in the Python code of your Lambda function
```

* As an environment variable under the name `DD_API_KEY`
* As a KMS Encrypted key under the name `DD_KMS_API_KEY`

```
module initialization error: The API key is not valid.
```

> The API key you provided doesn't match any key Datadog recognizes. Double-check the spelling and make sure that your key corresponds to the org you are sending data to. If your org is in Europe, you need to specify a `DD_SITE` variable with the value `datadoghq.eu` in order for the API key to be properly matched with your account.

Reviewer 3: ```suggestion

The API key you provided doesn't match any key Datadog recognizes. Double-check the spelling and make sure that your key corresponds to the org you are sending data to. If your are in Datadog EU site, you need to specify a `DD_SITE` variable with the value `datadoghq.eu` in order for the API key to be properly matched with your account.
```

```
module initialization error: The API key is not the expected length. Please confirm that your API key is correct
```

> The API key is either too short or too long. Double-check that you copied it over correctly.

Reviewer 3: ```suggestion

The API key is either too short or too long. Double-check that you copied it over correctly.
```

## Check for conflicting subscriptions

AWS doesn't allow for more than one resource to be subscribed to a log source. If you have triggers on your Lambda function but it still isn't being invoked, there may be a conflict with another resource already subscribed to the same log source. The log source needs to be freed up so the Datadog Lambda function can collect logs from it.

> 

Reviewer 2: A screenshot of a log group list that highlight the subscriber (last column on the right) could also be helpful to let users double check if there is not another function or service consuming the logs.

If something is already subscribing to a Log Group that you want to monitor with the Datadog Lambda Function, you can remove it: 

> * Select the log source 

Reviewer 3: ```suggestion

* Select the log source 
```

**Note** If you have triggers on your Lambda function but it still isn't being invoked, there may be a conflict with another resource already subscribed to the same log source. When you add a manual trigger, an error message informs you if a resource is already subscribed to the log source:

{{< img src="logs/guide/creating-trigger-error-example.png" alt="Example of error when creating trigger with subscription"  style="width:80%;" >}}

See [Check for conflicting subscriptions](#check-for-conflicting-subscriptions) for more information on removing subscriptions.

[1]: https://docs.datadoghq.com/integrations/amazon_web_services/?tab=allpermissions#manually-setup-triggers
{{% /tab %}}
{{% tab "Automatic trigger" %}}

You can see if there are [automatic triggers][1] set up for your Lambda function with the following steps:

1. Navigate to the [Cloudwatch console][2].
2. Click **Log Groups** in the left sidebar here you will see a list of Log Groups. There is a **subscriptions** column on the right that shows what resources (if any) are currently subscribed to the log source.

> {{< img src="logs/guide/log-group-subcriptions-example.png" alt="Log Group subcriptions example"  style="width:80%;" >}}

Reviewer 3: ```suggestion
{{< img src="logs/guide/log-group-subscriptions-example.png" alt="Log Group subcriptions example"  style="width:80%;" >}}
```
